src/reactor.cc

```diff
@@ -0,0 +1,406 @@
+      cyclus::Material::Ptr m = inventory.Pop();
+      base->Absorb(m);
+    }
+
+    inventory.Push(base);
+  }
+}
+
+void Reactor::ExtractHelium(
+    cyclus::toolkit::ResBuf<cyclus::Material>& inventory) {
+  if (!inventory.empty()) {
+    cyclus::Material::Ptr mat = inventory.Pop();
+    cyclus::CompMap c = mat->comp()->atom();
+    cyclus::compmath::Normalize(&c, mat->quantity());
+
+    cyclus::CompMap He3 = {{20030000, 1}};
+
+    // A threshold of 1e-5 was set to allow tritium_reserve inventories up to
+    // 1000kg. A 1 decade lower threshold prevents tritium_reserve inventories
+    // above 33kg.
+    cyclus::Material::Ptr helium = mat->ExtractComp(
+        c[20030000], cyclus::Composition::CreateFromAtom(He3), 1e-5);
+
+    helium_storage.Push(helium);
+    inventory.Push(mat);
+  }
+}
+
+void Reactor::RecordEvent(std::string name, std::string val) {
+  context()
+      ->NewDatum("ReactorEvents")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Event", name)
+      ->AddVal("Value", val)
+      ->Record();
+}
+
+void Reactor::RecordOperationalInfo(std::string name, std::string val) {
+  context()
+      ->NewDatum("ReactorOperationsLog")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Event", name)
+      ->AddVal("Value", val)
+      ->Record();
+}
+
+void Reactor::RecordStatus(std::string status, double power) {
+  context()
+      ->NewDatum("ReactorStatus")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("Status", status)
+      ->AddVal("Power", power)
+      ->Record();
+}
+
+void Reactor::RecordInventories(double core, double reserve, double storage,
+                                double blanket, double helium) {
+  context()
+      ->NewDatum("ReactorInventories")
+      ->AddVal("AgentId", id())
+      ->AddVal("Time", context()->time())
+      ->AddVal("TritiumCore", core)
+      ->AddVal("TritiumReserve", reserve)
+      ->AddVal("TritiumStorage", storage)
+      ->AddVal("LithiumBlanket", blanket)
+      ->AddVal("HeliumStorage", helium)
+      ->Record();
+}
+
+void Reactor::DepleteBlanket(double bred_tritium_mass) {
+  cyclus::Material::Ptr blanket_mat = blanket.Pop();
+
+  cyclus::CompMap b = blanket_mat->comp()->mass();
+  cyclus::compmath::Normalize(&b, blanket_mat->quantity());
+
+  // Percent (as decimal) of T which comes from Li-7 instead of Li-6
+  double Li7_contribution = 0.15;
+
+  cyclus::CompMap depleted_comp;
+
+  // This is ALMOST the correct behavior, but "scraping the bottom of the
+  // barrel" is a little too complex for this implementation.
+  if ((b[30060000] - (1 - Li7_contribution) * 2 * bred_tritium_mass > 0) &&
+      (b[30070000] - Li7_contribution * 7.0 / 3.0 * bred_tritium_mass > 0)) {
+    depleted_comp = {{30070000, b[30070000] - Li7_contribution * 7.0 / 3.0 *
+                                                  bred_tritium_mass},
```

Comment:
I could be wrong, but because this references b[30070000] and b[30060000] (the quantity of Li6 and Li7 in the blanket), and the blanket composition changes over time as it gets depleted and recycled, it is necessary to recalculate this value every time more tritium is bred. This is also why there's an if statement before it checking whether there's enough Li6 and Li7 left in the blanket to proceed with the depletion.